### PR TITLE
Added tiling to LAROMANCE model

### PR DESCRIPTION
**Pull request by tophmatthews**
This grew into a monster... It includes a total reorganization of the ROM computation to limit execution time, inclusion of ROM tiling, and better handling of input parameters outside the window of applicability. There will likely be some optimization of the methods, but this captures at least the verified answer for code-to-code comparison.

Ref #14046

### CODE REVIEW

**Comment by tophmatthews** [comment]
Ping @bwspenc and @sapitts for review. This is an update to the ROM. This lays some of the groundwork for some soon-to-follow patch-work ROM that will utilize several ROM's with unique input windows that will be combined to cover a large range of parameter space with small errors and small polynomial degrees.

File: modules/tensor_mechanics/include/materials/ADLAROMANCEStressUpdateBase.h
Comment (bwspenc) [inline modules/tensor_mechanics/include/materials/ADLAROMANCEStressUpdateBase.h:32]: Don't comment out the variable names in the header. You only need to do that in the .C file if they are unused.

File: modules/tensor_mechanics/include/materials/ADLAROMANCEStressUpdateBase.h
Comment (bwspenc) [inline modules/tensor_mechanics/include/materials/ADLAROMANCEStressUpdateBase.h:71]: Looks like most of the `@param` instances above need to be removed now.

File: modules/tensor_mechanics/include/materials/ADLAROMANCEStressUpdateBase.h
Comment (bwspenc) [inline modules/tensor_mechanics/include/materials/ADLAROMANCEStressUpdateBase.h:78]: ...and you need to add some here. Same goes for all the other functions with changed signatures.

File: modules/tensor_mechanics/include/materials/ADLAROMANCEStressUpdateBase.h
Comment (bwspenc) [inline modules/tensor_mechanics/include/materials/ADLAROMANCEStressUpdateBase.h:200]: ?

File: modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C
Comment (bwspenc) [inline modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:233]: Maybe it would be better to multiply by a number slightly less than one (so it is independent of the magnitude of stress)

File: modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C
Comment (bwspenc) [inline modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:333]: Seems odd to send this to the error stream (since it's not an error).

File: modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C
Comment (bwspenc) [inline modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:465]: Probably want to check for divide by zero here. Also, I'm not really well versed in AD -- does transform_coef need to be an ADReal, or is it not necessary because it wouldn't change, and therefore doesn't affect the Jacobian?

**Review by bwspenc — CHANGES_REQUESTED** [review]

File: modules/tensor_mechanics/include/materials/ADLAROMANCEStressUpdateBase.h
Comment (tophmatthews) [reply modules/tensor_mechanics/include/materials/ADLAROMANCEStressUpdateBase.h:208]: oh my, how embarrassing....

File: modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C
Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:340]: Yeah, this is to ensure that it is actually output when running in parallel. If it is just Moose::out and not on the main mpi process, it will be buried.

File: modules/tensor_mechanics/include/materials/ADLAROMANCEStressUpdateBase.h
Comment (tophmatthews) [reply modules/tensor_mechanics/include/materials/ADLAROMANCEStressUpdateBase.h:88]: ... and these

File: modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C
Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:235]: Do you mean slightly more than 0?

File: modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C
Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:472]: Yeah, exactly about ADReal/Real. I'll add a check for transform_coef in the initialize stage, good point.

File: modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C
Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:235]: Oh, no i see, you're right....do you think it matters?

Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:235]: Actually, yeah, if stress is really high, the 1e-9 will get lost in machine precision?

**Comment by tophmatthews** [comment]
Well, I'm not sure how you want to do this @bwspenc. I am working through some bigger changes to allow for ROM tiling. You can wait until then, or do this incremental review now? The newer version will mostly be adding in things. This PR is mostly changing what is in there now.

File: modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C
Comment (tophmatthews) [inline modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:282]: It's always weird that github does this...I just moved computeTimeStepLimit towards the bottom...

File: modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C
Comment (sapitts) [inline modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:244]: I agree that this negative trial stress error is quite aggravating (all four of my laromance simulations I'm currently running are complaining with that precise error); however, I am really not a fan of this magic number introduction. Why 6 digits of 9s instead of 5 or 7? Is there a better way to achieve the same goal?

File: modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C
Comment (sapitts) [inline modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:156]: Adding in the 'en' to give index wouldn't change the clang formatting on these lines and would make the code more readable

File: modules/tensor_mechanics/include/materials/ADLAROMANCEStressUpdateBase.h
Comment (sapitts) [inline modules/tensor_mechanics/include/materials/ADLAROMANCEStressUpdateBase.h:279]: I had thought that Aaron wanted these to be a vector /meta-string/ thingambober that would give the positions of the inputs and outputs. Did you discuss that option and elect this set of indices instead?

File: modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C
Comment (sapitts) [inline modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:124]: Switching these dislocation terms away from the mobile and immobile that I used in my original code and to the cell and wall dislocations that Laurent always uses in his code and presentations would be a good step for consistency

**Review by sapitts — COMMENTED** [review]
I can live with most of these changes, excepting the use of the "magic number" coefficient. Perhaps @bwspenc and/or @dschwen would be willing to suggest a different way to avoid the negative trial stress error

File: modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C
Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:244]: I am open to suggestions, what would you propose?

File: modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C
Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:156]: I suppose

File: modules/tensor_mechanics/include/materials/ADLAROMANCEStressUpdateBase.h
Comment (tophmatthews) [reply modules/tensor_mechanics/include/materials/ADLAROMANCEStressUpdateBase.h:279]: I'm not aware of that, maybe we can chat about that this afternoon? This should achieve the same purpose no?

**Comment by tophmatthews** [comment]
> Ping @bwspenc and @sapitts for review. This is an update to the ROM. This lays some of the groundwork for some soon-to-follow patch-work ROM that will utilize several ROM's with unique input windows that will be combined to cover a large range of parameter space with small errors and small polynomial degrees.

Thanks for the look here @sapitts, but I have a ton of rewrites here that include the tiling. It's probably waiting for that (forthcoming this week.) I'll include your suggestions.

File: modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C
Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:156]: Updated

File: modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C
Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:124]: Updated

**Comment by tophmatthews** [comment]
UGGG I don't know why these are failing!!
```
tensor_mechanics/test:rom_stress_update.rom/3d: /opt/civet/conda_builds/conda_envs/pr-15310/x86_64-conda_cos6-linux-gnu/include/c++/7.3.0/debug/vector:417:
tensor_mechanics/test:rom_stress_update.rom/3d: In function:
tensor_mechanics/test:rom_stress_update.rom/3d: Segmentation fault
```

**Comment by tophmatthews** [comment]
@bwspenc @sapitts Well this PR is ready to go, but it is failing for a reason I can't figure out in dbg. These pass fine on my mac...

**Comment by sapitts** [comment]
I've pulled your branch @tophmatthews but I don't have great hopes that I'm going to be much help

**Comment by tophmatthews** [comment]
Yeah I'll probably have to build it on our cluster to try and find the failure...these ones are the worst where they work on mac, but fail on linux

File: modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C
Comment (sapitts) [reply modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:316]: @lynnmunday and I talked with @bwspenc today about this negative trial stress error problem and have another solution for this error: Ben has a check and reset approach to make sure that the inelastic effective strain (scalar) does not exceed the total mechanical strain in each time step. I believe that @lynnmunday will push up a commit with this change today

Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:316]: Awesome!!

**Comment by sapitts** [comment]
@tophmatthews you mentioned that you were going to build this on your linux cluster to debug; any success with finding the issues of the debug failures there?

**Comment by tophmatthews** [comment]
> @tophmatthews you mentioned that you were going to build this on your linux cluster to debug; any success with finding the issues of the debug failures there?

Fair Point! I'll try and do that. I was hoping for a magic fix from you MOOSIAN experts :) I'll try again.

**Comment by sapitts** [comment]
@tophmatthews on the "easy" jacobian tester fails, AD mastermind @lindsayad once told me a passing rate of 1e-7 was quite acceptable. If that's still true, one can get around the default 1e-8 value by adding the following in the tests file:
```
    difference_tol = 5e-8  #difference = {test specific}
    ratio_tol = 5e-8    #ratio = {test specific}
```

**Comment by lindsayad** [comment]
@tophmatthews has often complained (complained may be a strong word) about our default values for those parameters. I'm still waiting for him to make a PR changing them :smile:

File: modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C
Comment (lynnmunday) [inline modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:748]: This function needs to be changed to something like this to make it pass the debug test.  ln 748 is changing the size of the size 3 std::vector you created in ln 742.  Then line 754 transformed_limits[t][i][j][2] is out of bounds.  



std::vector<std::vector<std::vector<std::vector<Real>>>>
ADLAROMANCEStressUpdateBase::getTransformedLimits(
    const std::vector<std::vector<std::vector<Real>>> limits)
{
  std::vector<std::vector<std::vector<std::vector<Real>>>> transformed_limits(
      _num_tiles,
      std::vector<std::vector<std::vector<Real>>>(
          _num_outputs, std::vector<std::vector<Real>>(_num_inputs, std::vector<Real>(3))));

  for (unsigned int t = 0; t < _num_tiles; ++t)
  {
    for (unsigned int i = 0; i < _num_outputs; ++i)
    {
      // transformed_limits[t][i] = limits[t];   //limits[t][j].size()=2 and I think that overwrites the size 3
                                                 // you put have on ln 742 so when you try to write to transformed_limits[t][i][j][2] it is out of bounds.
      for (unsigned int j = 0; j < _num_inputs; ++j)
      {
        std::vector<Real> temp_vec(3, 0.0);
        for (unsigned int k = 0; k < 2; ++k)
        {
          temp_vec[k] = limits[t][j][k];
          convertValue(temp_vec[k], _transform[t][i][j], _transform_coefs[t][i][j]);
        }
        temp_vec[2] = (temp_vec[1] - temp_vec[0]) / 2.0;
        transformed_limits[t][i][j] = temp_vec;
      }
    }
  }

  return transformed_limits;
}

File: modules/tensor_mechanics/include/materials/ADLAROMANCEStressUpdateBase.h
Comment (sapitts) [inline modules/tensor_mechanics/include/materials/ADLAROMANCEStressUpdateBase.h:19]: In working with Aaron on the previous P91 ROM, I learned that the windows for the Inverse transform are switched, e.g. the low limit of effective strain (0.0) became the high window limit (+1) and the high limit of effective strain (~0.02 as an example) became the low window limit (-1). 

We'd need to add a case check to your convertValues method call on line 752 of ADLAROMANCEStressUpdateBase.C to switch the window limits for the inverse case. If you want to drop the INVERSE case for now (it's not being used in any LAROMANCE models that I know of), and add it later after we have time to think through an appropriate regression test, that would be fine with me

Comment (tophmatthews) [reply modules/tensor_mechanics/include/materials/ADLAROMANCEStressUpdateBase.h:19]: Whoa, that is way confusing...I'll double check

File: modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C
Comment (sapitts) [inline modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:352]: I haven't added a bunch of print statements to this yet, but it seemed like this check wasn't being enforced for the dislocations in my simulation. Not worth holding up this merge request but something that I'd like to look into on the next iteration

File: modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C
Comment (sapitts) [inline modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:797]: Again for a later merge request, in the simulation I ran with the new P91 model the check didn't seem to make a difference here for the dislocation densities that were below the bounds. Definitely user error on my part with not setting all the flags could have been an issue.

That said, considering an exception to cut the time step or something similar if the dislocations are outside of the evolution ranges might be prudent here

Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:797]: did you set `cell_input_window_failure` to the default value of `error`? That may have been the problem if you were using an old version of the code?

File: modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C
Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:748]: nice find! How did you figure it out? Did it spit out an error when doing gdb in unix?

File: modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C
Comment (sapitts) [reply modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:797]: I did not; I just ran with the defaults. I will add that option into my input files. thanks

Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:797]: Well thats a bit werid since the default is already error though

Comment (sapitts) [reply modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:797]: well petunias

Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:797]: Try running with verbose=true to see if the global windows make sense

Comment (sapitts) [reply modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:797]: Oh, I am already running with that option, and yes, the global values make sense. I double checked those values against the values from the getInputLimits method and they matched. 

I'll push up the demo input file and move that discussion over there

File: modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C
Comment (lynnmunday) [reply modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C]: gdb on unix gave me a backtrace with the line number where you access transformed_limits[t][i][j][2]
maybe valgrind on my mac would have found it.

File: modules/tensor_mechanics/include/materials/ADLAROMANCEStressUpdateBase.h
Comment (tophmatthews) [reply modules/tensor_mechanics/include/materials/ADLAROMANCEStressUpdateBase.h:19]: I'm not too sure if this is right or not. Since we don't need it for any of the current implementations, I'll drop it for now. We can circle back if we need to.

**Comment by sapitts** [comment]
Solid mechanics strikes again! this time in app documentation files

**Comment by tophmatthews** [comment]
> Solid mechanics strikes again! this time in app documentation files

Not me right? :)

**Comment by tophmatthews** [comment]
I trust you'll restart the tests once the doco is fixed @sapitts ? Anything else I need to fix before then?

**Comment by sapitts** [comment]
@bwspenc this pull request has now made it successfully through testing. Would you take another look at it and merge this pull request if it looks acceptable to you?

File: modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C
Comment (bwspenc) [reply modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:315]: So, I guess we still need to find a better resolution to this, but for the time being, I think this is OK.

File: modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C
Comment (bwspenc) [inline modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:696]: We should at least explain what the magic numbers are in a future revision of this.

**Review by bwspenc — APPROVED** [review]
Looks good.

File: modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C
Comment (sapitts) [reply modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:315]: > So, I guess we still need to find a better resolution to this, but for the time being, I think this is OK.

I agree. I'm hopeful that with the checks on the dislocation densities, we could explore the check_ranges option that Lynn had tried again in the next development round

File: modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C
Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:696]: These numbers are used by Aaron in his ROM development...somewhat magical, but more tied into how those coefficients are created.

**Pull request merged**
